### PR TITLE
 Implement ravel_pytree for internal usage

### LINKED ISSUE #252: fori_collect does not work with non-float arrays
The issue is ravel_pytree which calls `vjp` and does not work with int32 or bool arrays. As such, `fori_collect` will fail if the default identity transform is used.

### PR DESCRIPTION
**Pull request by neerajprad**
Fixes #252, based on discussion in #266.

### CODE REVIEW

File: numpyro/util.py [hunk 1]
```diff
@@ -171,3 +172,27 @@ def decorator(destin_class):
...
+
+
+def _ravel_list(*leaves):
+    leaves_metadata = tree_map(lambda l: pytree_metadata(np.ravel(l), np.shape(l), np.size(l), lax.dtype(l)),
+                               leaves)
+    leaves_idx = np.cumsum(np.array((0,) + tuple(d.size for d in leaves_metadata)))
+
+    def unravel_list(arr): return [np.reshape(lax.dynamic_slice_in_dim(arr, leaves_idx[i], m.size),
+                                              m.shape).astype(m.dtype)
+                                   for i, m in enumerate(leaves_metadata)]
```
Comment (fehiepsi) [inline numpyro/util.py]: beautiful!!

File: numpyro/util.py [hunk 2]
```diff
@@ -171,3 +172,27 @@ def decorator(destin_class):
...
+
+pytree_metadata = namedtuple('pytree_metadata', ['flat', 'shape', 'size', 'dtype'])
+
+
+def _ravel_list(*leaves):
+    leaves_metadata = tree_map(lambda l: pytree_metadata(np.ravel(l), np.shape(l), np.size(l), lax.dtype(l)),
+                               leaves)
+    leaves_idx = np.cumsum(np.array((0,) + tuple(d.size for d in leaves_metadata)))
+
+    def unravel_list(arr): return [np.reshape(lax.dynamic_slice_in_dim(arr, leaves_idx[i], m.size),
```
Comment (fehiepsi) [inline numpyro/util.py]: nit: I think it is easier to read if we start with new line (or use `unravel_list = lambda arr: ...`)

File: numpyro/util.py [hunk 3]
```diff
@@ -171,3 +172,27 @@ def decorator(destin_class):
...
+                                   for i, m in enumerate(leaves_metadata)]
+
+    return np.concatenate([m.flat for m in leaves_metadata]), unravel_list
+
+
+def ravel_pytree(pytree):
+    leaves, treedef = tree_flatten(pytree)
+    flat, unravel_list = _ravel_list(*leaves)
+
+    def unravel_pytree(arr): return tree_unflatten(treedef, unravel_list(arr))
```
Comment (fehiepsi) [inline numpyro/util.py]: nit: having return statement in newline

**Comment by neerajprad** [comment]
> Could you also update fori_collect using this?

`fori_collect` will just pick it up since its defined in the same file, and we aren't importing `ravel_pytree` from jax in that file.

File: numpyro/util.py (hunk 3 above)
Comment (neerajprad) [reply numpyro/util.py]: Done.

File: numpyro/util.py (hunk 2 above)
Comment (neerajprad) [reply numpyro/util.py]: Done

**Pull request merged**